File: backend-kotlin/app/api/src/main/kotlin/com/wallet/clover/api/controller/NotificationController.kt
package com.wallet.clover.api.controller

import com.wallet.clover.api.common.CommonResponse
import com.wallet.clover.api.common.PageResponse
import com.wallet.clover.api.dto.Notification
import com.wallet.clover.api.service.NotificationService
import org.springframework.web.bind.annotation.*
import java.time.LocalDateTime

@RestController
@RequestMapping("/api/v1/notifications")
class NotificationController(
    private val notificationService: NotificationService
) {

    @GetMapping
    suspend fun getMyNotifications(
        @RequestParam userId: Long,
        @RequestParam(defaultValue = "0") page: Int,
        @RequestParam(defaultValue = "20") size: Int
    ): CommonResponse<PageResponse<Notification.Response>> {
        val notifications = notificationService.getMyNotifications(userId, page, size)
        val response = PageResponse(
            content = notifications.content.map { Notification.Response.from(it) },
            page = notifications.page,
            size = notifications.size,
            totalElements = notifications.totalElements,
            totalPages = notifications.totalPages
        )
        return CommonResponse.success(response)
    }

    @PostMapping
    suspend fun createNotification(
        @RequestParam userId: Long,
        @RequestParam title: String,
        @RequestParam message: String
    ): CommonResponse<String> {
        notificationService.createNotification(userId, title, message)
        return CommonResponse.success("Notification created")
    }

    @PutMapping("/{notificationId}/read")
    suspend fun markAsRead(
        @PathVariable notificationId: Long,
        @RequestParam userId: Long
    ): CommonResponse<String> {
        notificationService.markAsRead(notificationId, userId)
        return CommonResponse.success("Notification marked as read")
    }

    @GetMapping("/unread-count")
    suspend fun getUnreadCount(@RequestParam userId: Long): CommonResponse<Long> {
        return CommonResponse.success(notificationService.getUnreadCount(userId))
    }
}

File: apps/backend/src/notification/notification.controller.ts
import {
  Controller,
  Get,
  Post,
  Put,
  Body,
  Param,
  Query,
  UseGuards,
  Request,
  Req,
} from '@nestjs/common';
import { NotificationService } from './notification.service';
import { FcmService } from './fcm.service';
import { AuthGuard } from '@nestjs/passport';

@Controller('notifications')
export class NotificationController {
  constructor(
    private readonly notificationService: NotificationService,
    private readonly fcmService: FcmService,
  ) {}

  @UseGuards(AuthGuard('jwt'))
  @Get()
  async getMyNotifications(
    @Request() req: any,
    @Query('page') page = 0,
    @Query('size') size = 20,
  ) {
    return this.notificationService.getMyNotifications(
      req.user.id,
      +page,
      +size,
    );
  }

  @UseGuards(AuthGuard('jwt'))
  @Put(':id/read')
  async markAsRead(@Param('id') id: string, @Request() req: any) {
    await this.notificationService.markAsRead(BigInt(id), req.user.id);
    return { success: true };
  }

  @UseGuards(AuthGuard('jwt'))
  @Get('unread-count')
  async getUnreadCount(@Request() req: any) {
    return this.notificationService.getUnreadCount(req.user.id);
  }

  @Post()
  @UseGuards(AuthGuard('jwt'))
  async createNotification(
    @Req() req: any,
    @Body() body: { title: string; message: string; type?: string },
  ) {
    const userId = req.user.id;
    return this.notificationService.createNotification(
      userId,
      body.title,
      body.message,
      body.type || 'INFO',
    );
  }

  @UseGuards(AuthGuard('jwt'))
  @Post('fcm-token')
  async registerFcmToken(@Request() req: any, @Body() body: { token: string }) {
    await this.fcmService.registerTokenById(req.user.id, body.token);
    return { success: true };
  }
}
